packages/quantic/cypress/integration/facets/timeframe-facet/timeframe-facet-expectations.ts

```diff
@@ -0,0 +1,190 @@
+      selector
+        .expandButton()
+        .should(display ? 'exist' : 'not.exist')
+        .logDetail(`The expand button ${should(display)} be displayed.`),
+
+    displayPlaceholder: (display: boolean) =>
+      selector
+        .placeholder()
+        .should(display ? 'exist' : 'not.exist')
+        .logDetail(`The facet placeholder ${should(display)} be displayed.`),
+
+    valuesEqual: (values: string[]) =>
+      selector
+        .valueLabel()
+        .then((elements) => {
+          const actualValueLabels = Cypress.$.makeArray(elements).map(
+            (element) => element.innerText
+          );
+          expect(actualValueLabels).to.deep.equal(values);
+        })
+        .logDetail(
+          `The facet values should strictly equal: ${values
+            .map((v) => `"${v}"`)
+            .join(', ')}.`
+        ),
+
+    displayValues: (display: boolean) =>
+      selector
+        .values()
+        .should(display ? 'exist' : 'not.exist')
+        .logDetail(`The facet values ${should(display)} be displayed.`),
+
+    numberOfValues: (value: number) =>
+      selector
+        .values()
+        .should('have.length', value)
+        .logDetail(`The facet should have ${value} values.`),
+
+    displayClearButton: (display: boolean) =>
+      selector
+        .clearFilterButton()
+        .should(display ? 'exist' : 'not.exist')
+        .logDetail(`The clear filter button ${should(display)} be displayed.`),
+
+    urlHashContains: (field: string, range: string) => {
+      const expectedHash = `#df[${field}]=${range}`;
+      cy.url()
+        .should('include', expectedHash)
+        .logDetail(`The URL hash should contain "${expectedHash}"`);
+    },
+
+    urlHashIsEmpty: () =>
+      cy.url().should('match', /#?$/).logDetail('The URL hash should be empty'),
+  };
+}
+
+function timeframeFacetWithValuesExpectations(
+  selector: FacetWithValuesSelector
+) {
+  return {
+    numberOfSelectedValues: (value: number) =>
+      selector
+        .selectedValue()
+        .should('have.length', value)
+        .logDetail(`The facet should have ${value} selected values.`),
+
+    selectedValueContains: (value: string) =>
+      selector
+        .selectedValue()
+        .should('contain', value)
+        .logDetail(`The selected value should be "${value}".`),
+
+    logSelectedValue: (field: string, range: string) =>
+      cy
+        .wait(InterceptAliases.UA.Facet.Select)
+        .then((interceptor) => {
+          const analyticsBody = interceptor.request.body;
+
+          expect(analyticsBody).to.have.property('actionCause', 'facetSelect');
+          expect(analyticsBody).to.have.property('customData');
+          const customData = analyticsBody.customData;
+          expect(customData).to.have.property('facetField', field);
+          expect(customData).to.have.property('facetValue', range);
+        })
+        .logDetail('Should log the "facetSelect" UA event.'),
+
+    logClearFilter: (field: string) =>
```

Comment:
perhaps out of scope but these two log expectations seem similar enough and other tests probable have some that could also be replaced by some helper function with a signature something like this:
```
expectLogs: (alias: string, uaEvent: string, bodyData: {key: string, value: string}[], customData?: {key: string, value: string}[]) =>
      cy
        .wait(alias)
        .then((interceptor) => {
          const analyticsBody = interceptor.request.body;
          bodyData.forEach(entry => {
            expect(analyticsBody).to.have.property(entry.key, entry.value);
          })
          if (!customData) {
            return;
          }
          expect(analyticsBody).to.have.property('customData');
          customData.forEach(entry => {
              expect(analyticsBody.customData).to.have.property(entry.key, entry.value);
          });
        })
        .logDetail(`Should log the "${uaEvent}" UA event.`),
```

Reply:
You are right, the UA expectations are very similar and we can certainly use a generic function like you suggest.

Reply:
I modified the implementation a little to use the type

    Record<string, string | number | boolean>
    
instead of

    {key: string, value: string}[]

It made the code simpler when calling the common function.

Reply:
Looks great!